### ISSUE TITLE
Disk watermark validation rejects apparently valid sequence of updates

### ISSUE THREAD

**Issue by DaveCTurner** [elastic/elasticsearch#28309]
```
$ elasticsearch-6.1.1/bin/elasticsearch --version
Java HotSpot(TM) 64-Bit Server VM warning: Option UseConcMarkSweepGC was deprecated in version 9.0 and will likely be removed in a future release.
Version: 6.1.1, Build: bd92e7f/2017-12-17T20:23:25.338Z, JVM: 9.0.1
$ java -version
java version "9.0.1"
Java(TM) SE Runtime Environment (build 9.0.1+11)
Java HotSpot(TM) 64-Bit Server VM (build 9.0.1+11, mixed mode)
$ uname -a
Darwin Davids-MacBook-Pro.local 17.3.0 Darwin Kernel Version 17.3.0: Thu Nov  9 18:09:22 PST 2017; root:xnu-4570.31.3~1/RELEASE_X86_64 x86_64
```

On a fresh one-node cluster, the first of these operations succeeds but the second fails:
```
PUT _cluster/settings
{"transient":{"cluster.routing.allocation.disk.watermark.flood_stage":"99%"}}

PUT _cluster/settings
{"transient":{"cluster.routing.allocation.disk.watermark.high":"97%"}}
```

The result is:
```
{
  "error": {
    "root_cause": [
      {
        "type": "illegal_argument_exception",
        "reason": "high disk watermark [97%] more than flood stage disk watermark [95%]"
      }
    ],
    "type": "illegal_argument_exception",
    "reason": "high disk watermark [97%] more than flood stage disk watermark [95%]"
  },
  "status": 400
}
```

This is surprising because the flood stage watermark was set to `99%` but the error message claims it is still `95%`. As expected, `GET _cluster/settings` yields:

```
{
  "persistent": {},
  "transient": {
    "cluster": {
      "routing": {
        "allocation": {
          "disk": {
            "watermark": {
              "flood_stage": "99%"
            }
          }
        }
      }
    }
  }
}
```

Setting both watermarks together succeeds:

```
PUT _cluster/settings
{
  "transient": {
    "cluster.routing.allocation.disk.watermark.flood_stage": "99%",
    "cluster.routing.allocation.disk.watermark.high": "97%"
  }
}
```

I see the same effect replacing `transient` with `persistent` throughout too.

**Comment by dylanPowers** [comment]
This also happens with high versus low and is still present in v6.2.3
```
GET _cluster/settings
{
  "persistent": {
    "cluster": {
      "routing": {
        "allocation": {
          "disk": {
            "watermark": {
              "low": "50%"
            }
```
```
PUT _cluster/settings 
{
  "persistent": {
    "cluster.routing.allocation.disk.watermark.high":  "66%"
  }
}

Result:
{
  "error": {
    "root_cause": [
      {
        "type": "remote_transport_exception",
        "reason": "[tw-r03-s01.master-elasticsearch-logging-staging.service.chcg01.consul][172.23.73.194:9300][cluster:admin/settings/update]"
      }
    ],
    "type": "illegal_argument_exception",
    "reason": "low disk watermark [85%] more than high disk watermark [66%]"
  },
  "status": 400
}

**Comment by faxm0dem** [comment]
I can confirm this is still in 6.3.2 thanks to @Xylakant for pointing me here

**Comment by cbismuth** [comment]
Looks interesting, I'd like to work on it if no one has already started to do so.

**Comment by cbismuth** [comment]
From `master` branch it seems we can't set theses value with the percent unit anymore. I'll switch to branch 6.3 to reproduce this issue.

```
{
  "error": {
    "root_cause": [
      {
        "type": "illegal_argument_exception",
        "reason": "illegal value can't update [cluster.routing.allocation.disk.watermark.low] from [1b] to [1b]"
      }
    ],
    "type": "illegal_argument_exception",
    "reason": "illegal value can't update [cluster.routing.allocation.disk.watermark.low] from [1b] to [1b]",
    "caused_by": {
      "type": "illegal_argument_exception",
      "reason": "unable to consistently parse [cluster.routing.allocation.disk.watermark.low=1b], [cluster.routing.allocation.disk.watermark.high=1b], and [cluster.routing.allocation.disk.watermark.flood_stage=90%] as percentage or bytes",
      "caused_by": {
        "type": "parse_exception",
        "reason": "failed to parse setting [cluster.routing.allocation.disk.watermark.flood_stage] with value [90%] as a size in bytes: unit is missing or unrecognized"
      }
    },
    "suppressed": []
  },
  "status": 400
}
```

**Comment by cbismuth** [comment]
My bad, when sending all three `low`, `high` and `flood_stage` with consistent values in percent, it all works. I'll debug a bit deeper and let you know.

**Request**

```
{
  "transient": {
    "cluster.routing.allocation.disk.watermark.low": "50%",
    "cluster.routing.allocation.disk.watermark.high": "90%",
    "cluster.routing.allocation.disk.watermark.flood_stage": "95%"
  }
}
```

**Response**

```
{
  "acknowledged": true,
  "persistent": {},
  "transient": {
    "cluster": {
      "routing": {
        "allocation": {
          "disk": {
            "watermark": {
              "low": "50%",
              "flood_stage": "95%",
              "high": "90%"
            }
          }
        }
      }
    }
  }
}
```

**Comment by cbismuth** [comment]
Alright, I finally have a complete cURL recreation script, see Gist [here](https://gist.github.com/cbismuth/295c14974fbbe962b7a850aae2aeabdf). Debug time!

**Comment by cbismuth** [comment]
Issue found and reproduced on `master` branch:
* It happens that when setting only the `high` value, it's validated against settings built from constants with default values.

I'm working on a fix.

**References**

Here is below the line in which we're reading settings for validation (first line).
https://github.com/elastic/elasticsearch/blob/040695b64e6452c9e54109b9d31fe3d1efdd69f7/server/src/main/java/org/elasticsearch/common/settings/Setting.java#L406-L417

Here is below how this settings are built from previous method call.
https://github.com/elastic/elasticsearch/blob/31aabe4bf9ea73cbb1c21322d9e9aa2a578b41a0/server/src/main/java/org/elasticsearch/cluster/routing/allocation/DiskThresholdSettings.java#L104-L110

Ratio validation is implemented here.
https://github.com/elastic/elasticsearch/blob/31aabe4bf9ea73cbb1c21322d9e9aa2a578b41a0/server/src/main/java/org/elasticsearch/cluster/routing/allocation/DiskThresholdSettings.java#L174-L186

**Comment by cbismuth** [comment]
@jasontedor @rjernst @s1monw 

I've reworked my comments on this issue to make them easier to read. Could you please have a look at them and tell me what do you think? Thanks a lot.

**Comment by DaveCTurner** [comment]
Hi @cbismuth, you can reproduce this in a unit test as follows:

```diff
diff --git a/server/src/test/java/org/elasticsearch/cluster/routing/allocation/DiskThresholdSettingsTests.java b/server/src/test/java/org/elasticsearch/cluster/routing/allocation/DiskThresholdSettingsTests.java
index 342fcea7dde..b99c13e2d99 100644
--- a/server/src/test/java/org/elasticsearch/cluster/routing/allocation/DiskThresholdSettingsTests.java
+++ b/server/src/test/java/org/elasticsearch/cluster/routing/allocation/DiskThresholdSettingsTests.java
@@ -203,4 +203,10 @@ public class DiskThresholdSettingsTests extends ESTestCase {
         assertThat(cause, hasToString(containsString("low disk watermark [85%] more than high disk watermark [75%]")));
     }

+    public void testSequenceOfUpdates() {
+        final ClusterSettings clusterSettings = new ClusterSettings(Settings.EMPTY, ClusterSettings.BUILT_IN_CLUSTER_SETTINGS);
+        new DiskThresholdSettings(Settings.EMPTY, clusterSettings); // this has the effect of registering the settings updater
+        clusterSettings.applySettings(Settings.builder().put(DiskThresholdSettings.CLUSTER_ROUTING_ALLOCATION_DISK_FLOOD_STAGE_WATERMARK_SETTING.getKey(), "99%").build());
+        clusterSettings.applySettings(Settings.builder().put(DiskThresholdSettings.CLUSTER_ROUTING_ALLOCATION_HIGH_DISK_WATERMARK_SETTING.getKey(), "97%").build());
+    }
 }
```

I'm not sure what you mean by "these dependencies aren't expressed in the code" - e.g. the `LowDiskWatermarkValidator` declares its dependency on the `high` and `flood_stage` settings here:

https://github.com/elastic/elasticsearch/blob/31aabe4bf9ea73cbb1c21322d9e9aa2a578b41a0/server/src/main/java/org/elasticsearch/cluster/routing/allocation/DiskThresholdSettings.java#L105-L110

**Comment by cbismuth** [comment]
Sorry, I'll reword. You're right @DaveCTurner, static dependencies are expressed in the code. But when triggering validation, runtime cluster state values aren't passed to the validator. Therefore, the validator validates input against default values.

In my previous comment, the _runtime cluster state values_ are enclosed in the `target` parameter of `AbstractScopedSettings#updateSettings` API. And as you can see in this snippet I referenced, this parameter isn't passed to `AbstractScopedSettings#validate` validation API.

Is it clearer?

**Comment by cbismuth** [comment]
Yes, this API has a lot of usages. Thank you for the unit test, I’ll play with it :+1:

**Comment by cbismuth** [comment]
I'll try next week to add the `target` parameter of the `updateSettings` API (L702) as a third parameter to the `validate` API (L719), with a `Settings.EMPTY` default value everywhere else.

I'll see how big are the changes and how it may break things. I'll let you know.

**Comment by cbismuth** [comment]
I had a look to available validation APIs in `AbstractScopedSettings` class, and I've opened PR https://github.com/elastic/elasticsearch/pull/34184 to suggest an enriched version of the actual validation.

No test seems to be broken on my laptop.

There may be a small overhead, but I let you give me you opinion on this.

The  bug reported by @DaveCTurner is fixed with these changes, and here is a sample output when `high` is higher than `flood_stage`.

```
{
  "error" : {
    "root_cause" : [
      {
        "type" : "illegal_argument_exception",
        "reason" : "high disk watermark [98%] more than flood stage disk watermark [97%]"
      }
    ],
    "type" : "illegal_argument_exception",
    "reason" : "high disk watermark [98%] more than flood stage disk watermark [97%]",
    "suppressed" : [
      {
        "type" : "illegal_argument_exception",
        "reason" : "high disk watermark [98%] more than flood stage disk watermark [97%]"
      },
      {
        "type" : "illegal_argument_exception",
        "reason" : "high disk watermark [98%] more than flood stage disk watermark [97%]"
      }
    ]
  },
  "status" : 400
}
```

**Comment by cbismuth** [comment]
@DaveCTurner I've added the unit test you suggested.
I had to slightly modify it by [reusing returned settings](https://github.com/cbismuth/elasticsearch/blob/71cfe1a874d550056d8b7facbc9b7acb04b92779/server/src/test/java/org/elasticsearch/cluster/routing/allocation/DiskThresholdSettingsTests.java#L217) from the `AbstractScopedSettings#applySettings` API. Otherwise we couldn't have kept track of updated settings in a unit test.

Does it make sense to you or did I miss something?

**Comment by DaveCTurner** [comment]
Closed by #34184.
